src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java

```diff
@@ -68,9 +70,27 @@ protected JmxProxy connect(Optional<RepairStatusHandler> handler, String host, i
       password = jmxAuth.getPassword();
     }
     try {
+      if (JMX_CONNECTIONS.containsKey(host) && JMX_CONNECTIONS.get(host).isConnectionAlive()) {
+        return JMX_CONNECTIONS.get(host);
+      }
+
       JmxProxy jmxProxy = JmxProxyImpl.connect(handler, host, username, password, addressTranslator, connectionTimeout);
       hostConnectionCounters.incrementSuccessfulConnections(host);
-      return jmxProxy;
+      JmxProxy previousConnection = JMX_CONNECTIONS.putIfAbsent(host, jmxProxy);
+      if (previousConnection == null) {
+        // The putIfAbsent succeeded and our JmxProxy object is the one in the map
+        LOG.info("Adding new JMX Proxy for host {}", host);
+
+      } else {
+        if (previousConnection.isConnectionAlive()) {
+          LOG.info("Reusing previously created JMX connection for host {}", host);
+          jmxProxy.close();
+        } else {
+          LOG.info("Recreating JMX connection for host {}", host);
```

Comment:
the previousConnection should be closed here

```
previousConnection.close();
```

Reply:
I've reworked that part to make it safer and avoid creating a new JMX connection only to close it if another one already existed.
The new code involves a synchronized section to make sure we're not re-creating the connections several times and to simplify the code (contention is not an issue here as the synchronized block is only used if there's no existing connection for that host).